dbms/src/DataStreams/MergingAndConvertingBlockInputStream.h

```diff
@@ -141,31 +156,45 @@ class MergingAndConvertingBlockInputStream : public IProfilingBlockInputStream
                 if (parallel_merge_data->exception)
                     std::rethrow_exception(parallel_merge_data->exception);
 
-                auto it = parallel_merge_data->ready_blocks.find(current_bucket_num);
-                if (it != parallel_merge_data->ready_blocks.end())
+
+                if (!parallel_merge_data->ready_blocks.empty())
                 {
-                    ++current_bucket_num;
-                    scheduleThreadForNextBucket();
+                    size_t finished_thread_num = parallel_merge_data->ready_blocks.size();
+
+                    for (size_t i = 0; i < finished_thread_num; ++i)
+                    {
+                        scheduleThreadForNextBucket();
+                    }
 
-                    if (it->second)
+                    for (auto & ready_block : parallel_merge_data->ready_blocks)
                     {
-                        res.swap(it->second);
-                        break;
+                        current_bucket_num++;
+                        if (!ready_block.second.empty())
+                        {
+                            two_level_blocks.splice(two_level_blocks.end(), std::move(ready_block.second), ready_block.second.begin(), ready_block.second.end());
+                        }
+                    }
+                    parallel_merge_data->ready_blocks.clear();
+
+                    if (Block out = tryGetTwoLevelOutputBlock())
```

Comment:
I think even if `parallel_merge_data->ready_blocks.empty()` is true, we may still get a non-empty block from `tryGetTwoLevelOutputBlock()`, so I think we need to call `tryGetTwoLevelOutputBlock` outside the if branch again?

Reply:
And please double check if we also need to check `current_bucket_num` outside the if branch.

Reply:
My mistake, you already call `tryGetTwoLevelOutputBlock` before entering the while loop.